# Data exploratory

In [1]:
import json
def read_data(path):
    # read meta data
    with open(path, 'r') as fp:
        products = json.load(fp)
        return products

In [2]:
import numpy as np
import pandas as pd

products_json = read_data("../data/meta.json")
# initilize object arrays
product_id = np.empty((len(products_json['meta'])), dtype=object)
brand = np.empty((len(products_json['meta'])), dtype=object)
category = np.empty((len(products_json['meta'])), dtype=object)
subcategory = np.empty((len(products_json['meta'])), dtype=object)
product_name = np.empty((len(products_json['meta'])), dtype=object)

# collect values to numpy array
for index in range(len(products_json['meta'])):
    product_id[index] = products_json['meta'][index]['productid']
    brand[index] = products_json['meta'][index]['brand']
    category[index] = products_json['meta'][index]['category']
    subcategory[index] = products_json['meta'][index]['subcategory']
    product_name[index] = products_json['meta'][index]['name']

# write data into pandas DataFrame
products = pd.DataFrame()
products['productid'] = product_id
products['brand'] = brand
products['category'] = category
products['subcategory'] = subcategory
products['productname'] = product_name
del product_id, brand, category, subcategory, product_name
products

productid        brand              category  \
0        HBV00000AX6LR      Palette         Kişisel Bakım   
1        HBV00000BSAQG         Best              Pet Shop   
2        HBV00000JUHBA  Tarım Kredi            Temel Gıda   
3        HBV00000NE0QI        Namet  Et, Balık, Şarküteri   
4        HBV00000NE0UQ     Muratbey    Kahvaltılık ve Süt   
...                ...          ...                   ...   
10231  ZYHPREISBBKL008         Reis            Temel Gıda   
10232    ZYNES11470137      Nescafe             İçecekler   
10233   ZYPAREX1909309        Parex  Ev Bakım ve Temizlik   
10234   ZYPAREX2107986        Parex  Ev Bakım ve Temizlik   
10235       ZYPYON6690      Febreze  Ev Bakım ve Temizlik   

                          subcategory  \
0                          Saç Bakımı   
1                                Kedi   
2           Bakliyat, Pirinç, Makarna   
3                           Şarküteri   
4                              Peynir   
...                               ...   
10231       Bakliyat, Pirinç, Makarna   
10232                           Kahve   
10233            Ev Temizlik Ürünleri   
10234            Yüzey Temizleyiciler   
10235  Oda Kokusu ve Koku Gidericiler   

                                             productname  
0              Palette Kalıcı Doğal Renkler 10-4 PAPATYA  
1      Best Pet Jöle İçinde Parça Etli Somonlu Konser...  
2           Türkiye Tarım Kredi Koop.Yeşil Mercimek 1 kg  
3                      Namet Fıstıklı Macar Salam 100 gr  
4                          Muratbey Burgu Peyniri 250 gr  
...                                                  ...  
10231                       Reis Gönen Baldo Pirinç 1 kg  
10232   Nescafé Classic Çözünebilir Kahve 200gr Ekopaket  
10233                   Parex Powerball Çizmez Ovma Teli  
10234         Parex Tornado Döner Başlıklı Temizlik Seti  
10235  Febreze Kumaş ve Çamaşır Ferahlatıcı Sprey 500...  

[10236 rows x 5 columns]

In [3]:
products.isnull().any()

productid      True
brand          True
category       True
subcategory    True
productname    True
dtype: bool

In [4]:
products[products['productid'].isna()]

productid brand category subcategory productname
5092      None  None     None        None        None

In [5]:
products[products['category'].isna()]

productid brand category subcategory productname
5092      None  None     None        None        None

In [6]:
products[products['subcategory'].isna()]

productid brand category subcategory productname
5092      None  None     None        None        None

In [7]:
products[products['productname'].isna()]

productid brand category subcategory productname
5092      None  None     None        None        None

In [8]:
products.iloc[5092]

productid      None
brand          None
category       None
subcategory    None
productname    None
Name: 5092, dtype: object

In [9]:
# remove 5092 index because all values are null
products.drop([5092], inplace=True)
products.isnull().any()

productid      False
brand           True
category       False
subcategory    False
productname    False
dtype: bool

In [10]:
# process for evetns.json
events_json = read_data("../data/events.json")

event = np.empty((len(events_json['events'])), dtype=object)
sessionid = np.empty((len(events_json['events'])), dtype=object)
eventtime = np.empty((len(events_json['events'])), dtype=object)
price = np.empty((len(events_json['events'])), dtype=object)
productid = np.empty((len(events_json['events'])), dtype=object)

for index in range(len(events_json['events'])):
    event[index] = events_json['events'][index]['event']
    sessionid[index] = events_json['events'][index]['sessionid']
    eventtime[index] = events_json['events'][index]['eventtime']
    price[index] = events_json['events'][index]['price']
    productid[index] = events_json['events'][index]['productid']

events = pd.DataFrame()
events['event'] = event
events['sessionid'] = sessionid
events['eventtime'] = eventtime
events['price'] = price
events['productid'] = productid
del event, sessionid, eventtime, price, productid
events

event                             sessionid                 eventtime  \
0       cart  a0655eee-1267-4820-af21-ad8ac068ff7a  2020-06-01T08:59:16.406Z   
1       cart  d2ea7bd3-9235-4a9f-a9ea-d7f296e71318  2020-06-01T08:59:46.580Z   
2       cart  5e594788-78a0-44dd-8e66-37022d48f691  2020-06-01T08:59:33.308Z   
3       cart  fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf  2020-06-01T08:59:31.911Z   
4       cart  9e9d4f7e-898c-40fb-aae9-256c40779933  2020-06-01T08:59:33.888Z   
...      ...                                   ...                       ...   
387651  cart  0ce386d3-f9ae-46b1-a018-baa5ea91be0c  2020-06-15T03:55:05.322Z   
387652  cart  69be07e4-737d-4935-b2eb-e02ecd94986b  2020-06-15T03:56:50.064Z   
387653  cart  25f6960c-0144-419d-8761-2b539c75c9c9  2020-06-15T03:03:23.859Z   
387654  cart  eebd42a5-2d95-41b9-bc10-83db56ba3989  2020-06-15T03:29:54.865Z   
387655  cart  69be07e4-737d-4935-b2eb-e02ecd94986b  2020-06-15T03:55:04.997Z   

        price      productid  
0       14.48  HBV00000NVZE8  
1       49.90  HBV00000U2B18  
2        1.99   OFIS3101-080  
3        2.25  HBV00000NVZBW  
4        9.95  HBV00000NE0T4  
...       ...            ...  
387651  19.90  HBV00000OE884  
387652   9.95  HBV00000OE8AU  
387653  21.95  HBV00000PQM1C  
387654  17.50  HBV00000NFHXT  
387655  27.99  HBV00000O2S4F  

[387656 rows x 5 columns]

In [11]:
events.isnull().any()

event        False
sessionid    False
eventtime    False
price         True
productid     True
dtype: bool

In [12]:
events[events['price'].isna()]

event                             sessionid                 eventtime  \
6421    cart  fc1ba51a-224b-4818-bff0-fa2dc1c93acd  2020-06-01T12:32:22.012Z   
86682   cart  967f0239-a7f0-4ca9-b31e-92ad4e0798d7  2020-06-01T21:41:43.474Z   
193736  cart  78ca9e26-f4ee-4fb3-a166-b124fdc32426  2020-06-03T21:17:37.145Z   
197084  cart  25f19afd-6cb2-46dd-9ebd-0fd97cf10b63  2020-06-03T19:03:31.293Z   
343563  cart  cd86710a-f292-49e7-b48e-6b57d51755ae  2020-06-06T21:27:48.620Z   
354628  cart  1e50875d-c395-4e6d-8ea4-52e2bcb4975a  2020-06-04T06:08:54.849Z   

       price productid  
6421    None      None  
86682   None      None  
193736  None      None  
197084  None      None  
343563  None      None  
354628  None      None

In [13]:
events[events['productid'].isna()]

event                             sessionid                 eventtime  \
6421    cart  fc1ba51a-224b-4818-bff0-fa2dc1c93acd  2020-06-01T12:32:22.012Z   
86682   cart  967f0239-a7f0-4ca9-b31e-92ad4e0798d7  2020-06-01T21:41:43.474Z   
193736  cart  78ca9e26-f4ee-4fb3-a166-b124fdc32426  2020-06-03T21:17:37.145Z   
197084  cart  25f19afd-6cb2-46dd-9ebd-0fd97cf10b63  2020-06-03T19:03:31.293Z   
343563  cart  cd86710a-f292-49e7-b48e-6b57d51755ae  2020-06-06T21:27:48.620Z   
354628  cart  1e50875d-c395-4e6d-8ea4-52e2bcb4975a  2020-06-04T06:08:54.849Z   

       price productid  
6421    None      None  
86682   None      None  
193736  None      None  
197084  None      None  
343563  None      None  
354628  None      None

In [14]:
# check which session has how many products
events.groupby(["sessionid"]).agg({"productid": 'count'}).reset_index()

sessionid  productid
0      000280f4-62fc-4dcd-b51d-c66ac14d7d8c          1
1      0002e53b-1f60-4309-8380-31ca03de51f8          2
2      0002ef34-6bee-4953-874b-8298ec26b625          2
3      000618de-d415-408c-863e-6124db43f529          4
4      000770d6-c2d4-4ad2-bb2c-b35274bc5e7e          1
...                                     ...        ...
54437  fffb5e6a-2676-4cd9-b4e4-ab8b6621e0fe          6
54438  fffbae5f-8102-4a14-84d7-6c11c724cf8d          2
54439  fffbba74-6999-460f-bd5f-70eaebe689cf         20
54440  fffd3c61-2f71-4437-986c-e1c30ef5b5fe         43
54441  ffffcd3c-da03-4667-9c75-9fcafb609c9e         13

[54442 rows x 2 columns]

In [15]:
# detect how many null values exist
events.isna().value_counts()

event  sessionid  eventtime  price  productid
False  False      False      False  False        387650
                             True   True              6
dtype: int64

# Data preparation

In [16]:
# drop columns
events = events.drop(columns=['event', 'eventtime', 'price'])
events

sessionid      productid
0       a0655eee-1267-4820-af21-ad8ac068ff7a  HBV00000NVZE8
1       d2ea7bd3-9235-4a9f-a9ea-d7f296e71318  HBV00000U2B18
2       5e594788-78a0-44dd-8e66-37022d48f691   OFIS3101-080
3       fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf  HBV00000NVZBW
4       9e9d4f7e-898c-40fb-aae9-256c40779933  HBV00000NE0T4
...                                      ...            ...
387651  0ce386d3-f9ae-46b1-a018-baa5ea91be0c  HBV00000OE884
387652  69be07e4-737d-4935-b2eb-e02ecd94986b  HBV00000OE8AU
387653  25f6960c-0144-419d-8761-2b539c75c9c9  HBV00000PQM1C
387654  eebd42a5-2d95-41b9-bc10-83db56ba3989  HBV00000NFHXT
387655  69be07e4-737d-4935-b2eb-e02ecd94986b  HBV00000O2S4F

[387656 rows x 2 columns]

Collaborative Filter tries to detect similar users and recommends products. For more information you can watch https://www.youtube.com/watch?v=9AP-DgFBNP4&ab_channel=ArtificialIntelligence-AllinOne
or you can just read on Wikipedia https://en.wikipedia.org/wiki/Collaborative_filtering

The below code creates a zeros numpy array which size is 54442 × 10236. These numbers comes from dataset. There are 54442 sessions and 10236 unit products. 

In [18]:
sessions = events.sessionid.unique()
productids = events.productid.unique()
matrix = np.zeros(shape=(sessions.shape[0], productids.shape[0]))
for session_step, ses in enumerate(sessions):
    session_products = events[events["sessionid"] == ses].productid.unique()
    for p in session_products:
        idx = np.where(productids == p)
        matrix[session_step][idx] = 1
matrix

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [20]:
# Here we prepera the dataset
dataset = pd.DataFrame(data=matrix, columns=productids, index=sessions, dtype=int)
dataset

HBV00000NVZE8  HBV00000U2B18  \
a0655eee-1267-4820-af21-ad8ac068ff7a              1              0   
d2ea7bd3-9235-4a9f-a9ea-d7f296e71318              0              1   
5e594788-78a0-44dd-8e66-37022d48f691              0              0   
fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf              0              0   
9e9d4f7e-898c-40fb-aae9-256c40779933              0              0   
...                                             ...            ...   
f5668d65-04d7-4553-b42d-cd2be8c3a5c4              0              0   
7c5361ba-255b-4d5c-8443-8ada934a760d              0              0   
86663b4d-827e-4d23-9949-31750c17b7cc              0              0   
8fc5e99c-3655-45b2-99a9-47f7ff229fde              0              0   
d784e0c9-4b64-42e0-8c2a-74cb290a26d2              0              0   

                                      OFIS3101-080  HBV00000NVZBW  \
a0655eee-1267-4820-af21-ad8ac068ff7a             0              0   
d2ea7bd3-9235-4a9f-a9ea-d7f296e71318             0              0   
5e594788-78a0-44dd-8e66-37022d48f691             1              0   
fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf             0              1   
9e9d4f7e-898c-40fb-aae9-256c40779933             0              0   
...                                            ...            ...   
f5668d65-04d7-4553-b42d-cd2be8c3a5c4             0              0   
7c5361ba-255b-4d5c-8443-8ada934a760d             0              0   
86663b4d-827e-4d23-9949-31750c17b7cc             0              0   
8fc5e99c-3655-45b2-99a9-47f7ff229fde             0              0   
d784e0c9-4b64-42e0-8c2a-74cb290a26d2             0              0   

                                      HBV00000NE0T4  HBV00000NE0T6  \
a0655eee-1267-4820-af21-ad8ac068ff7a              0              0   
d2ea7bd3-9235-4a9f-a9ea-d7f296e71318              0              0   
5e594788-78a0-44dd-8e66-37022d48f691              0              0   
fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf              0              0   
9e9d4f7e-898c-40fb-aae9-256c40779933              1              1   
...                                             ...            ...   
f5668d65-04d7-4553-b42d-cd2be8c3a5c4              0              0   
7c5361ba-255b-4d5c-8443-8ada934a760d              0              0   
86663b4d-827e-4d23-9949-31750c17b7cc              0              0   
8fc5e99c-3655-45b2-99a9-47f7ff229fde              0              0   
d784e0c9-4b64-42e0-8c2a-74cb290a26d2              0              0   

                                      HBV00000O2SDO  HBV00000U2B4I  \
a0655eee-1267-4820-af21-ad8ac068ff7a              0              0   
d2ea7bd3-9235-4a9f-a9ea-d7f296e71318              0              0   
5e594788-78a0-44dd-8e66-37022d48f691              0              0   
fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf              0              0   
9e9d4f7e-898c-40fb-aae9-256c40779933              0              0   
...                                             ...            ...   
f5668d65-04d7-4553-b42d-cd2be8c3a5c4              0              0   
7c5361ba-255b-4d5c-8443-8ada934a760d              0              0   
86663b4d-827e-4d23-9949-31750c17b7cc              0              0   
8fc5e99c-3655-45b2-99a9-47f7ff229fde              0              0   
d784e0c9-4b64-42e0-8c2a-74cb290a26d2              0              0   

                                      HBV00000NVZBI  HBV000008SJOK  ...  \
a0655eee-1267-4820-af21-ad8ac068ff7a              0              0  ...   
d2ea7bd3-9235-4a9f-a9ea-d7f296e71318              0              0  ...   
5e594788-78a0-44dd-8e66-37022d48f691              0              0  ...   
fdfeb652-22fa-4153-b9b5-4dfa0dcaffdf              1              0  ...   
9e9d4f7e-898c-40fb-aae9-256c40779933              1              0  ...   
...                                             ...            ...  ...   
f5668d65-04d7-4553-b42d-cd2be8c3a5c4              0              0  ...   
7c5361ba-255b-4d5c-8443-8ada934a760d              0              0  ...   
86663b4d-82

In [21]:
# lets check the dataset size
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 54442 entries, a0655eee-1267-4820-af21-ad8ac068ff7a to d784e0c9-4b64-42e0-8c2a-74cb290a26d2
Columns: 10236 entries, HBV00000NVZE8 to HBV00000PV7M3
dtypes: int64(10236)
memory usage: 4.2+ GB


That dataset contain 0 and 1. So the data can be bool which mean the dataset size will be much lower. But Calculation is not possible with boolean data. Collaborative filtering need cosine similarity calculation. Use np.int8 instate of use int32

In [29]:
# Here we prepera the dataset
dataset = pd.DataFrame(data=matrix, columns=productids, index=sessions, dtype=np.int8)
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 54442 entries, a0655eee-1267-4820-af21-ad8ac068ff7a to d784e0c9-4b64-42e0-8c2a-74cb290a26d2
Columns: 10236 entries, HBV00000NVZE8 to HBV00000PV7M3
dtypes: int8(10236)
memory usage: 531.9+ MB


As above can seen, allocated much less memory. That means calculation will be much more efficient and same accurate

In [30]:
dataset.to_pickle("../data/cf_data.pkl")